lib/std/heap.zig

```diff
@@ -8,6 +8,378 @@ const c = std.c;
+        .x86, .x86_64 => 4 << 10,
+        .xtensa => 4 << 10,
+        else => null,
+    },
+    .freestanding => switch (builtin.cpu.arch) {
+        .wasm32, .wasm64 => 64 << 10,
+        else => null,
+    },
+    else => null,
+};
+
+/// The compile-time minimum page size that the target might have.
+/// All pointers from `mmap` or `VirtualAlloc` are aligned to at least `min_page_size`, but their
+/// actual alignment may be much bigger.
+/// This value can be overridden via `std.options.min_page_size`.
+/// On many systems, the actual page size can only be determined at runtime with `pageSize()`.
+pub const min_page_size: usize = std.options.min_page_size orelse (default_min_page_size orelse if (builtin.os.tag == .freestanding or builtin.os.tag == .other)
+    @compileError("freestanding/other explicitly has no min_page_size. One can be provided with std.options.min_page_size")
+else
+    @compileError(@tagName(builtin.cpu.arch) ++ "-" ++ @tagName(builtin.os.tag) ++ " has no min_page_size. One can be provided with std.options.min_page_size"));
+
+/// The compile-time maximum page size that the target might have.
+/// Targeting a system with a larger page size may require overriding `std.options.max_page_size`,
+/// as well as using the linker arugment `-z max-page-size=`.
+/// The actual page size can only be determined at runtime with `pageSize()`.
+pub const max_page_size: usize = std.options.max_page_size orelse (default_max_page_size orelse if (builtin.os.tag == .freestanding or builtin.os.tag == .other)
+    @compileError("freestanding/other explicitly has no max_page_size. One can be provided with std.options.max_page_size")
+else
+    @compileError(@tagName(builtin.cpu.arch) ++ "-" ++ @tagName(builtin.os.tag) ++ " has no max_page_size. One can be provided with std.options.max_page_size"));
+
+/// Returns the system page size.
+/// If the page size is comptime-known, `pageSize()` returns it directly.
+/// Otherwise, `pageSize()` defers to `std.options.queryPageSizeFn()`.
+pub fn pageSize() usize {
+    if (min_page_size == max_page_size) {
+        return min_page_size;
+    }
+    return std.options.queryPageSizeFn();
+}
+
+// A cache used by `defaultQueryPageSizeFn()` to avoid repeating syscalls.
+var page_size_cache = std.atomic.Value(usize).init(0);
+
+// The default implementation in `std.options.queryPageSizeFn`.
+// The first time it is called, it asserts that the page size is within the comptime bounds.
+pub fn defaultQueryPageSizeFn() usize {
+    var size = page_size_cache.load(.unordered);
+    if (size > 0) return size;
+    size = switch (builtin.os.tag) {
+        .linux => if (builtin.link_libc) @intCast(std.c.sysconf(@intFromEnum(std.c._SC.PAGESIZE))) else std.os.linux.getauxval(std.elf.AT_PAGESZ),
+        .driverkit, .ios, .macos, .tvos, .visionos, .watchos => blk: {
+            if (!builtin.link_libc)
+                @compileError("querying page size on Darwin is not supported without linking libc");
```

Comment:
You can drop this check since macOS is a "requires libc" platform in general, i.e. the compiler enforces it.